aiowebostv/webos_client.py

```diff
@@ -133,238 +135,258 @@ async def close_client_session(self) -> None:
+        Open additional connection needed to send button commands
+        The url is dynamically generated and returned from the ep.INPUT_SOCKET
+        endpoint on the main connection.
+        """
+        sockres = await self.request(ep.INPUT_SOCKET)
+        inputsockpath = sockres["socketPath"]
+        return await self._ws_connect(inputsockpath)
 
-            await asyncio.wait(handler_tasks, return_when=asyncio.FIRST_COMPLETED)
+    async def _get_states_and_subscribe_state_updates(self) -> None:
+        """Get initial states and subscribe to state updates.
 
-        except Exception as ex:  # pylint: disable=broad-except
-            if isinstance(ex, TimeoutError):
-                _LOGGER.debug("timeout(%s): connection", self.host)
-            else:
-                _LOGGER.debug("exception(%s): %r", self.host, ex, exc_info=True)
-            if not res.done():
-                res.set_exception(ex)
-        finally:
-            for task in handler_tasks:
-                if not task.done():
-                    task.cancel()
-
-            for future in self.futures.values():
-                future.cancel()
-
-            closeout = set()
-            closeout.update(handler_tasks)
-
-            if main_ws is not None:
-                closeout.add(asyncio.create_task(main_ws.close()))
-            if input_ws is not None:
-                closeout.add(asyncio.create_task(input_ws.close()))
-            if self.created_client_session:
-                closeout.add(asyncio.create_task(self.close_client_session()))
-
-            self.connection = None
-            self.input_connection = None
-
-            self.do_state_update = False
-
-            self._power_state = {}
-            self._current_app_id = None
-            self._muted = None
-            self._volume = None
-            self._current_channel = None
-            self._channel_info = None
-            self._channels = None
-            self._apps = {}
-            self._extinputs = {}
-            self._system_info = {}
-            self._software_info = {}
-            self._hello_info = {}
-            self._sound_output = None
-            self._media_state = []
+        Avoid partial updates during initial subscription.
+        """
+        self.do_state_update = False
+        self._system_info, self._software_info = await asyncio.gather(
+            self.get_system_info(), self.get_software_info()
+        )
+        subscribe_state_updates = {
+            self.subscribe_power_state(self.set_power_state),
+            self.subscribe_current_app(self.set_current_app_state),
+            self.subscribe_muted(self.set_muted_state),
+            self.subscribe_volume(self.set_volume_state),
+            self.subscribe_apps(self.set_apps_state),
+            self.subscribe_inputs(self.set_inputs_state),
+            self.subscribe_sound_output(self.set_sound_output_state),
+            self.subscribe_media_foreground_app(self.set_media_state),
+        }
```

Comment:
Is this actually faster, or does the cost of making more connections create more overhead vs using http keepalive?

Reply:
Nevermind on this, I was assuming `.request` was an HTTP request not a websocket request so there is no new connection overhead. I'm so use to `get` and `post` wrapping a function called `request` that I connected dots that were not here.